In [22]:
import numpy as np
import pandas as pd
# from aocd import get_data
import os
import re
import time

module_path=os.path.abspath(os.path.join('../'))
sys.path.append(module_path)
from ComplexMap import ComplexMap

# file="test1.txt"
# file="test2.txt"
file ="input.txt"

with open(file) as f:
    lines = f.readlines()
    
lines = list(map(lambda x:x.strip(),lines))

In [23]:
plan = ComplexMap(lines,str) 
plan_dict = plan.map_as_dict()
plan_set = plan.map_as_set()

In [24]:
def findAreaFromPoint(coord,val,processed):
    area = set()
    to_check = [coord]
    processed_copy = processed.copy()
    
    while to_check:
        el = to_check.pop(0)
        # print(">>",to_check)
        if el not in processed and not el in area and plan_dict[el]==val:
            area.add(el)
            to_check+=plan.get_neighbours(el)
    
    processed_copy.update(area)
    
    return area, processed_copy
            

In [25]:
def getArea(point_set):
    return len(point_set)

def getPerimeter_asSet(point_set):
    sides = set()
    for p in point_set:
        n_direct = plan.get_neighbour_directions()
        for d in n_direct:
            if p+d not in point_set:
                sides.add((p,d))

    return sides

def getPerimeter(point_set):
    return len(getPerimeter_asSet(point_set))

def getSides(point_set):
    sidesPotential = getPerimeter_asSet(point_set)
    sides = set()
    
    while sidesPotential:
        # print(sidesPotential)
        p1,d1 = sidesPotential.pop()
        orthDir = d1*1j
        # print(p1,d1,orthDir)
        i=1
        while (p1+i*orthDir,d1) in sidesPotential:
            # remove that new point from sides potential
            sidesPotential.remove((p1+i*orthDir,d1))
            i+=1
        i=1
        while (p1-i*orthDir,d1) in sidesPotential:
            # remove that new point from sides potential
            sidesPotential.remove((p1-i*orthDir,d1))
            i+=1

        sides.add((p1,d1))
    
    return len(sides)

def score_partA(point_set):
    area = getArea(point_set)
    per = getPerimeter(point_set)
    return area*per

def score_partB(point_set):
    area = getArea(point_set)
    nsides = getSides(point_set)
    return area*nsides


In [28]:
processed = set()
areas = {}


for coord, val in plan_set:
    if coord not in processed:
        #do BFS, will update processed
        area_set,processed = findAreaFromPoint(coord,val,processed)
        areas[coord]=area_set
        
# print(areas)

partA=0
partB=0
for a in areas:
    # print(score(areas[a]))
    partA+=score_partA(areas[a])
    partB+=score_partB(areas[a])
print("PartA",partA)
print("PartB",partB)

        

PartA 1456082
PartB 872382


test1:
PartA 140
PartB 80

test2:
PartA 772
PartB 436

input:
PartA 1456082
PartB 872382